pallets/manta-pay/src/lib.rs

```diff
@@ -480,19 +465,140 @@ pub mod pallet {
+
+		/// Maximum Size of Sender Data Update
+		const PULL_MAX_SENDER_UPDATE_SIZE: usize = 1024;
+
+		/// Pulls receiver data from the ledger starting at the `receiver_index`.
 		#[inline]
-		fn ledger() -> Ledger<T> {
-			Ledger(PhantomData)
+		fn pull_receivers(receiver_index: &mut [usize; 256]) -> (bool, ReceiverChunk) {
+			let mut more_receivers = false;
+			let mut receivers = Vec::new();
+			for (i, index) in receiver_index.iter_mut().enumerate() {
+				more_receivers |= Self::pull_receivers_for_shard(i as u8, index, &mut receivers);
+			}
+			(more_receivers, receivers)
+		}
+
+		/// Pulls receiver data from the shard at `shard_index` starting at the `receiver_index`,
+		/// pushing the results back to `receivers`.
+		#[inline]
+		fn pull_receivers_for_shard(
+			shard_index: u8,
+			receiver_index: &mut usize,
+			receivers: &mut ReceiverChunk,
+		) -> bool {
+			let mut iter = if *receiver_index == 0 {
+				Shards::<T>::iter_prefix(shard_index)
+			} else {
+				let raw_key = Shards::<T>::hashed_key_for(shard_index, *receiver_index as u64 - 1);
+				Shards::<T>::iter_prefix_from(shard_index, raw_key)
+			};
+			for _ in 0..Self::PULL_MAX_PER_SHARD_UPDATE_SIZE {
+				match iter.next() {
+					Some((_, (utxo, encrypted_note))) => {
+						*receiver_index += 1;
+						receivers.push((utxo, encrypted_note));
+					}
+					_ => return false,
+				}
+			}
+			iter.next().is_some()
+		}
+
+		/// Pulls sender data from the ledger starting at the `sender_index`.
+		#[inline]
+		fn pull_senders(sender_index: &mut usize) -> (bool, SenderChunk) {
+			let mut senders = Vec::new();
+			let mut iter = InsertionOrderedVoidNumbers::<T>::iter().skip(*sender_index);
+			for _ in 0..Self::PULL_MAX_SENDER_UPDATE_SIZE {
+				match iter.next() {
+					Some((_, sender)) => {
+						*sender_index += 1;
+						senders.push(sender);
+					}
+					_ => return (false, senders),
+				}
+			}
+			(iter.next().is_some(), senders)
 		}
 
-		/// The account ID of AssetManager
+		/// Returns the update required to be synchronized with the ledger starting from
+		/// `checkpoint`.
+		#[inline]
+		pub fn pull(mut checkpoint: Checkpoint) -> PullResponse {
+			let (more_receivers, receivers) = Self::pull_receivers(&mut checkpoint.receiver_index);
+			let (more_senders, senders) = Self::pull_senders(&mut checkpoint.sender_index);
+			PullResponse {
+				should_continue: more_receivers || more_senders,
+				checkpoint,
+				receivers,
+				senders,
+			}
+		}
+
+		/// Returns the account ID of this pallet.
+		#[inline]
 		pub fn account_id() -> T::AccountId {
 			T::PalletId::get().into_account()
 		}
+
+		/// Posts the transaction encoded in `post` to the ledger, using `sources` and `sinks` as
+		/// the public deposit and public withdraw accounts respectively.
+		#[inline]
+		fn post_transaction(
+			origin: Option<T::AccountId>,
+			sources: Vec<T::AccountId>,
+			sinks: Vec<T::AccountId>,
+			post: TransferPost,
+		) -> DispatchResultWithPostInfo {
+			Self::deposit_event(
```

Comment:
Did post_transaction go through some precheck before arriving here?
If not, we don't want to call deposit_event() in the failure case, as spamming the node with invalid txs could be a DDoS vector, as every event carries overhead.

Normal modus operandi is to deposit an event **on success only**

Reply:
Reading the code more carefully. The error path will jump out so no event will be emitted. 
Notice the two `?`